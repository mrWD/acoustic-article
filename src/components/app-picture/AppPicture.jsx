import React from 'react';
import PropTypes from 'prop-types';

export const getImgUrl = url => `${process.env.REACT_APP_DESTINATION_URL}${url}`;

const AppPicture = props => {
  if (!props.leadImage || (!props.leadImage.url && !props.leadImage.renditions)) {
    return null;
  }

  const src = getImgUrl(props.leadImage.url);
  let imgSourceList = null;

  if (props.leadImage.renditions) {
    imgSourceList = Object.values(props.leadImage.renditions)
      .sort((item1, item2) => item1.width - item2.width)
      .map((item) => (
        <source
          media={`(max-width: ${item.width}px)`}
          srcSet={getImgUrl(item.url)}
          key={item.url}
        ></source>
      ));
  }

  return (
    <picture>
      {imgSourceList}

      <img
        className={props.className}
        src={src}
        alt={props.altText}
      />
    </picture>
  );
};

AppPicture.propTypes = {
  className: PropTypes.string,
  altText: PropTypes.string,
  leadImage: PropTypes.shape({
    url: PropTypes.string,
    renditions: PropTypes.objectOf(
      PropTypes.shape({
        width: PropTypes.number.isRequired,
        url: PropTypes.string.isRequired,
      }),
    ),
  }),
};

AppPicture.defaultProps = {
  altText: '',
};

export default AppPicture;
