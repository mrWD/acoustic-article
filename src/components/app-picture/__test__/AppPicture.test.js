import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';

import AppPicture, { getImgUrl } from '../AppPicture';

configure({ adapter: new Adapter() });

describe('AppPicture', () => {
  test('renders without crashing', () => {
    shallow(<AppPicture />);
  });

  test('returns empty object, if leadImage is undefined, boolean or object without url and renditions', () => {
    const wrapperWithNull = shallow(<AppPicture />);
    const wrapperWithEmptyObject = shallow(<AppPicture leadImage={{}} />);

    expect(wrapperWithNull).toEqual({});
    expect(wrapperWithEmptyObject).toEqual({});
  });

  test('renders content without source tags, if leadImage does not have renditions', () => {
    const leadImage = { url: '/img-url.jpeg' };
    const wrapper = shallow(<AppPicture leadImage={leadImage} />);

    expect(wrapper.find('source')).toHaveLength(0);
    expect(wrapper).toMatchSnapshot();
  });

  test('renders content with source tags, if leadImage has renditions', () => {
    const leadImage = {
      url: '/img-url.jpeg',
      renditions: {
        default: {
          width: 1600,
          url: '/img-url.jpeg',
        },
        card: {
          width: 800,
          url: '/img-url-800.jpeg',
        },
      },
    };

    const wrapper = shallow(<AppPicture leadImage={leadImage} />);

    expect(wrapper.find('source')).toHaveLength(2);
    expect(wrapper).toMatchSnapshot();
  });
});

describe('getImgUrl', () => {
  test('returns process.env.REACT_APP_DESTINATION_URL + passed url', () => {
    const result = getImgUrl('/img-url.jpeg');

    expect(result).toBe(`${process.env.REACT_APP_DESTINATION_URL}/img-url.jpeg`);
  });
});
