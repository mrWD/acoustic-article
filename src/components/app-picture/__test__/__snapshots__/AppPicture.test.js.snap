// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppPicture renders content with source tags, if leadImage has renditions 1`] = `
<picture>
  <source
    key="/img-url-800.jpeg"
    media="(max-width: 800px)"
    srcSet="https://content-eu-4.content-cms.com/img-url-800.jpeg"
  />
  <source
    key="/img-url.jpeg"
    media="(max-width: 1600px)"
    srcSet="https://content-eu-4.content-cms.com/img-url.jpeg"
  />
  <img
    alt=""
    src="https://content-eu-4.content-cms.com/img-url.jpeg"
  />
</picture>
`;

exports[`AppPicture renders content without source tags, if leadImage does not have renditions 1`] = `
<picture>
  <img
    alt=""
    src="https://content-eu-4.content-cms.com/img-url.jpeg"
  />
</picture>
`;
